### PR TITLE
Create google maps api based shape factory

### PR DESCRIPTION
**Pull request by glasierr**

### CODE REVIEW

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    public GoogleMapsShapeFactory(String apiKey) {
+        this.geoApiContext = new GeoApiContext().setApiKey(apiKey);
+    }
+
+    @Override
+    public Shape create(List<Location> points) {
+        val resultElements = calculateDistanceMatrix(notEmpty(points)).rows[0].elements;
+
+        val shapePoints = new ArrayList<ShapePoint>();
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: new ArrayList<ShapePoint>(resultElements.length)

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/TransitRuntimeException.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package eu.socialedge.hermes.backend.core;
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/TransitRuntimeException.java]: Header missed

File: backend-core/src/test/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package eu.socialedge.hermes.backend.core;
```
Comment (wingsofovnia) [inline backend-core/src/test/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactoryTest.java:15]: header

File: dependencies.gradle [hunk 4]
```diff
@@ -40,6 +41,7 @@ ext {
             h2Driver: "com.h2database:h2:${versions.persistence.h2Driver}"
         ],
         geo: [
+            googleMapsApi: "com.google.maps:google-maps-services:${versions.geo.googleMapsApi}",
             unitsMeasurement: "tec.uom:uom-se:${versions.geo.unitsMeasurement}"
```
Comment (wingsofovnia) [inline dependencies.gradle]: it's not geo, but commons

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            shapePoints.add(new ShapePoint(pointLocation, quantity));
+        }
+
+        return new Shape(shapePoints);
+    }
+
+    private DistanceMatrix calculateDistanceMatrix(List<Location> points) {
+        val origin = toLatLng(points.get(0));
+        val destinations = points.stream()
+            .map(GoogleMapsShapeFactory::toLatLng)
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: skip(1) missed?

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            }
+
+            val quantity = Quantities.getQuantity(element.distance.inMeters, Units.METRE);
+            shapePoints.add(new ShapePoint(pointLocation, quantity));
+        }
+
+        return new Shape(shapePoints);
+    }
+
+    private DistanceMatrix calculateDistanceMatrix(List<Location> points) {
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: Consider this signature `Map<Location, Quantity> calculateDistanceTraveled(Location origin, List<Location> waypoints)`

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            .map(GoogleMapsShapeFactory::toLatLng)
+            .toArray(LatLng[]::new);
+
+        try {
+            return DistanceMatrixApi.newRequest(geoApiContext)
+                .origins(origin)
+                .destinations(destinations)
+                .await();
+        } catch (Exception e) {
+            throw new TransitRuntimeException("Exception occurred during distance matrix calculation", e);
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: It's not a TransitRuntimeException, but Http exception or IOException

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    public Shape create(List<Location> points) {
+        val resultElements = calculateDistanceMatrix(notEmpty(points)).rows[0].elements;
+
+        val shapePoints = new ArrayList<ShapePoint>();
+        for (int i = 0; i < resultElements.length; i++) {
+            val element = resultElements[i];
+            val pointLocation = points.get(i);
+
+            if (element.status != DistanceMatrixElementStatus.OK) {
+                throw new TransitRuntimeException("Couldn't recognize location " + pointLocation);
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: Imho all this factory has nothing to do with "transit". com.google.maps.errors.NotFoundException or generic "ShapeFactoryException" are a good candidates.

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            }
+
+            val quantity = Quantities.getQuantity(element.distance.inMeters, Units.METRE);
```
Comment (wingsofovnia) [inline backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: What was the case for previous pull request if you still give no choice for programmer who uses this factory? Consider adding Units to ShapeFactory#create:
```
Shape create(List<Location> points, Units units);
default Shape create(List<Location> points) {
   return create(points, Units.METRE);
}
```

**Review by wingsofovnia — CHANGES_REQUESTED** [review]

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java (hunk 9 above)
Comment (glasierr) [reply backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: The reason to add Quantity - to make it clear what unit of lenght is used and to make it easy for consumers to make conversions

File: backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java (hunk 5 above)
Comment (glasierr) [reply backend-core/src/main/java/eu/socialedge/hermes/backend/core/GoogleMapsShapeFactory.java]: No, we definetely need to have distance first - first point (will always be 0), according to gtfs specification

File: dependencies.gradle (hunk 4 above)
Comment (glasierr) [reply dependencies.gradle]: Wasn't it approved by you in previous pull request?

**Review by wingsofovnia — APPROVED** [review]

**Pull request merged**